### PR TITLE
Accessing Experiment by Name: updating service.proto

### PR DESCRIPTION
**Pull request by cryptexis**
As Id and name fields are unique for an experiment, this minor change adds the possibility to access experiment by its name. 
Tackling the #127

### CODE REVIEW

File: mlflow/protos/service.proto
Comment (mparkhe) [inline mlflow/protos/service.proto]: ... identifies the experiment.

File: mlflow/protos/service.proto
Comment (mparkhe) [inline mlflow/protos/service.proto]: My recommendation is to add a new API called `GetExperimentByName`, rather than overloading existing API.

**Review by mparkhe — DISMISSED** [review]

**Comment by cryptexis** [comment]
@mparkhe I have tried sharing the response message as you've suggested. After fighting couple of hours it  did not work out :( At this moment GetExperimentByName message has a same structure as GetExperiment. Would it be possible to return to  message definition after the rest of the functionality is implemented ? I have got the  whole flow working on my local branch. Can't wait to share, the solution :))

File: mlflow/protos/service.proto
Comment (mparkhe) [inline mlflow/protos/service.proto]: *nit* : indent

File: mlflow/protos/service.proto
Comment (mparkhe) [inline mlflow/protos/service.proto]: Can we reuse the `GetExperiment.Response` object?

File: mlflow/protos/service.proto
Comment (cryptexis) [reply mlflow/protos/service.proto]: Sure thing, done :)

File: mlflow/protos/service.proto
Comment (cryptexis) [reply mlflow/protos/service.proto]: I have tried couple of heuristics  and none of them seemed to work.  I  have tried the way it was proposed in the link you've provided:
https://developers.google.com/protocol-buffers/docs/proto#nested - didn't work so far.
It seems that GetExperimentByName explicitly requires to have a nested message object Response. I am not sure whether it is possible to do. Maybe you have some ideas?

File: mlflow/protos/databricks.proto
Comment (smurching) [inline mlflow/protos/databricks.proto]: Q: why this change?

File: tests/tracking/test_tracking.py
Comment (smurching) [inline tests/tracking/test_tracking.py]: Nit: we can now use the `tracking_uri_mock` to set our tracking URI to a tempdir, e.g.:

```
from tests.projects.utils import tracking_uri_mock
def test_get_experiment_by_name(tracking_uri_mock):
   # Do stuff, the tracking URI will be mocked to point to a temporary directory
```

File: mlflow/protos/service.proto
Comment (smurching) [inline mlflow/protos/service.proto]: Nit: this endpoint name sounds like we're getting the name of the experiment, maybe we could do `/preview/mlflow/experiments/get-by-name`?

**Review by smurching — COMMENTED** [review]
Thanks for the PR @cryptexis, this looks pretty good /  leaving a few questions - would you be able to rebase on master? Thanks :)

File: mlflow/protos/databricks.proto
Comment (cryptexis) [reply mlflow/protos/databricks.proto]: Hi :) 
Otherwise we get this warning....only for that reason
[libprotobuf WARNING google/protobuf/compiler/parser.cc:562] No syntax specified for the proto file: databricks.proto. Please use 'syntax = "proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted to proto2 syntax.)

File: mlflow/protos/service.proto
Comment (cryptexis) [reply mlflow/protos/service.proto]: fixed :)

File: tests/tracking/test_tracking.py
Comment (cryptexis) [reply tests/tracking/test_tracking.py]: I have looked up other examples using tracking_uri_mock. It was a bit unclear. If you don't mind I have used the same structure as the unit test for creating experiment.

**Comment by cryptexis** [comment]
@smurching I think, it's ready :)
I will add comments to the issue with regards to the changes in API :)

File: mlflow/store/rest_store.py
Comment (smurching) [inline mlflow/store/rest_store.py]: Let's hold off on adding experiment deletion APIs in this PR - @mparkhe is working on experiment deletion if you're interested in collaborating / reviewing :)

File: mlflow/store/file_store.py
Comment (smurching) [inline mlflow/store/file_store.py]: Maybe raise `ValueError` instead of the base `Exception`?

File: tests/tracking/test_tracking.py
Comment (smurching) [inline tests/tracking/test_tracking.py]: If we're not using `tracking_uri_mock`, could we store the result of `tempfile.mkdtemp` and remove it in our `finally` block to avoid leaking tempdirs? Alternatively, I think it wouldn't be too bad to use `tracking_uri_mock` here - it'd allow us to not worry about manipulating the tracking URI/cleaning up the temp directory, and we'd basically just rewrite the test as:

```
def test_get_experiment_by_name(tracking_uri_mock):

    service = get_service()

    with pytest.raises(TypeError):
        service.get_experiment_by_name()

    with pytest.raises(Exception):
        service.get_experiment_by_name("")

    with pytest.raises(Exception):
        service.get_experiment_by_name(None)

    service = get_service()
    rand_name = "Some random experiment name %d" % random.randint(1, 1e6)
    exp_id = mlflow.create_experiment(rand_name)
    assert exp_id is not None
    experiment = service.get_experiment_by_name(rand_name)
    assert experiment.experiment_id == exp_id
```

I'd prefer to use `tracking_uri_mock`, but also ok to leave this as is & if we clean up the tempdirs.

File: mlflow/protos/service.proto
Comment (smurching) [inline mlflow/protos/service.proto]: Could you add a brief comment header like:
```
// Gets the experiment with the specified name.
```
above this line? The comment will show up in autogenerated REST API docs / make it easier to use the API.

**Review by smurching — COMMENTED** [review]
Thanks @cryptexis, this is looking pretty good - just a couple small comments, once you address them / rebase on master this LGTM!

**Comment by smurching** [comment]
Also sorry the protos keep getting updated under this PR, they've been changing rapidly recently.

**Comment by cryptexis** [comment]
@smurching all is fixed as you requested. s3ArtifactRepo tests do not pass. Although I have noticed it's a common issue for now :)

**Review by smurching — APPROVED** [review]
LGTM! #376 should fix the S3 test failure, will merge after that's in - thanks @cryptexis!

**Comment by cryptexis** [comment]
Hi @smurching, I was wondering if you still find this functionality useful ?

**Comment by smurching** [comment]
@cryptexis ack really sorry for the delay...this functionality is definitely useful. Sorry, this is entirely my fault - but after conversation with some other committers, it seems the best way to implement this for now might be to just add logic in the rest_store similar to the file_store ([link](https://github.com/mlflow/mlflow/blob/master/mlflow/store/file_store.py#L161)) that calls `list_experiments` & then searches through the result. It's inefficient, but we can consider adding an endpoint that directly fetches the experiment by name later if/as performance becomes a bottleneck. 

Since you've already worked really hard to get this PR to the current state / had to rebase multiple times, I'm happy to resolve the merge conflicts & update the implementation. It'd be great to get a review from you, and your unit tests will be really helpful in verifying correctness of the implementation. Also happy to review & merge the necessary changes if you want to make them. Sorry again for the miscommunication & thanks for contributing!!

**Comment by cryptexis** [comment]
@smurching  Sure, no problem :) let's  get this done :))

**Comment by smurching** [comment]
Great! Just to confirm, do you plan to update the PR, or should I? (no rush!)

**Comment by jmrr** [comment]
Hey guys, did a new PR get created? I couldn't find it nor saw a change in this one's implementation. FWIW I arrived here after I also implemented myself some quick logic similar to that in the `file_store` as @smurching pointed out, so there's definitely demand for the feature. 

Thanks a lot for your work, this is going to be a great tool in the Data Science world.

**Comment by smurching** [comment]
@jmrr good question, #592 adds `get_experiment_by_name` support to the RestStore. I'll go ahead & close this PR for now - thanks again @cryptexis for working on this.

**Pull request closed without merging**
